clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -0,0 +1,341 @@
+    }
+
+    private NetworkClientDelegate.UnsentRequest makeHeartbeatRequest() {
+        // TODO: We only need to send the rebalanceTimeoutMs field once unless the first request failed.
+        ConsumerGroupHeartbeatRequestData data = new ConsumerGroupHeartbeatRequestData()
+            .setGroupId(membershipManager.groupId())
+            .setMemberEpoch(membershipManager.memberEpoch())
+            .setMemberId(membershipManager.memberId())
+            .setRebalanceTimeoutMs(rebalanceTimeoutMs);
+
+        membershipManager.groupInstanceId().ifPresent(data::setInstanceId);
+
+        if (this.subscriptions.hasPatternSubscription()) {
+            // TODO: Pass the string to the GC if server side regex is used.
+        } else {
+            data.setSubscribedTopicNames(new ArrayList<>(this.subscriptions.subscription()));
+        }
+
+        this.membershipManager.assignorSelection().serverAssignor().ifPresent(data::setServerAssignor);
+
+        NetworkClientDelegate.UnsentRequest request = new NetworkClientDelegate.UnsentRequest(
+            new ConsumerGroupHeartbeatRequest.Builder(data),
+            coordinatorRequestManager.coordinator());
+
+        request.future().whenComplete((response, exception) -> {
+            if (exception == null) {
+                onResponse((ConsumerGroupHeartbeatResponse) response.responseBody(), response.receivedTimeMs());
+            } else {
+                onFailure(exception, response.receivedTimeMs());
+            }
+        });
+        return request;
+    }
+
+    private void onFailure(final Throwable exception, final long responseTimeMs) {
+        this.heartbeatRequestState.onFailedAttempt(responseTimeMs);
+        logger.warn("Failed to send heartbeat to coordinator node {} due to error: {}",
+                coordinatorRequestManager.coordinator(), exception.getMessage());
+        if (exception instanceof RetriableException) {
+            logger.debug("Retrying heartbeat request in {}ms due to {}",
+                heartbeatRequestState.remainingBackoffMs(responseTimeMs),
+                exception.getMessage());
+        } else {
+            logger.error("Heartbeat request failed due to fatal error", exception);
+            membershipManager.transitionToFailed();
+            nonRetriableErrorHandler.handle(exception);
+        }
+    }
+
+    private void onResponse(final ConsumerGroupHeartbeatResponse response, long currentTimeMs) {
+        if (Errors.forCode(response.data().errorCode()) == Errors.NONE) {
+            this.heartbeatRequestState.updateHeartbeatIntervalMs(response.data().heartbeatIntervalMs());
+            this.heartbeatRequestState.onSuccessfulAttempt(currentTimeMs);
+            this.heartbeatRequestState.resetTimer();
+            this.membershipManager.updateState(response.data());
+            return;
+        }
+        onErrorResponse(response, currentTimeMs);
+    }
+
+    private void onErrorResponse(final ConsumerGroupHeartbeatResponse response,
```

Comment:
I expect this will be the place where we'll need to make sure to release any assignment we may have, right? Whenever a transitionToFailed/Fence occurs, the HB manager should call the reconciler to trigger the onPartitionsLost/Revoked as needed. Let's add a TODO just as we did for the missing bit of triggering the HB before the interval.